README.rst
Introduction
============

.. image:: https://badge.fury.io/py/bottle-rest.png
    :target: https://pypi.python.org/pypi/bottle-rest

.. image:: https://img.shields.io/pypi/dm/bottle-rest.svg
    :target: https://pypi.python.org/pypi/bottle-rest

.. image:: https://readthedocs.org/projects/bottle-rest/badge/?version=latest
    :target: http://bottle-rest.readthedocs.org/

.. image:: https://img.shields.io/github/issues/Bystroushaak/bottle-rest.svg
    :target: https://github.com/Bystroushaak/bottle-rest/issues

.. image:: https://img.shields.io/pypi/l/bottle-rest.svg

This package is used to make easier creating REST applications using
`Bottle <http://bottlepy.org>`_ web framework.

Documentation
-------------

Full module documentation and description can be found at Read the Docs:

- http://bottle-rest.readthedocs.org/


CHANGES.rst
Changelog
=========

0.6.0
-----
    - Added `@get_to_params`.
    - Code reformatted.
    - Removed support for py2.7.

0.5.2
-----
    - Change in the JSON imports that makes it easier to override the json encoder/decoder, if necessary. (by https://github.com/eevleevs in #11)

0.5.1
-----
    - Provided default for raw_data (by https://github.com/eevleevs in #10)

0.5.0
-----
    - Added support for python 3 (#9).
    - Added support for file-like objects (#7).

0.4.3
-----
    - Fixed separator in JSON prettify.

0.4.2
-----
    - Fixed #8 - deprecation warning when accessing the Exception.message property.

0.4.1
-----
    - Merged #5 to fix ``docs/__init__.py``, which didn't work at Python 3.
    - Rewritten ``docs/__init__.py``, when the merge for Python 3 broken Python 2 support.

0.4.0
-----
    - Added ``pretty_dump()`` decorator.

0.3.2
-----
    - Fixed bug in @wraps, caused by some strange python behaviour.

0.3.1
-----
    - Fixed bug in documentation.

0.3.0
-----
    - As suggested in #3 - project renamed from ``rbottle`` to ``bottle-rest``.

0.2.0
-----
    - Fixed some bugs, added optional return_json=True parameters.
    - Added pretty print for returned JSON.

0.1.0
-----
    - Project created.


run_tests.sh
#!/usr/bin/env sh

export PYTHONPATH="src:$PYTHONPATH"
python2 -m pytest unittests $@
python3 -m pytest unittests $@


unittests/test_rbottle.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
import sys
from StringIO import StringIO

import pytest

sys.path = ['src/rbottle'] + sys.path
import rbottle
from rbottle import HTTPError


# Functions & classes =========================================================
class MockRequest:
    def __init__(self, json):
        self.body = StringIO(json)
        self.forms = {}


def test_decode_json_body():
    rbottle.request = MockRequest('{"hello": "there"}')

    assert rbottle.decode_json_body() == {"hello": "there"}

    with pytest.raises(HTTPError):
        rbottle.request = MockRequest('{')
        rbottle.decode_json_body()


def test_encode_json_body():
    assert rbottle.encode_json_body({"one": 1}) == '''{
    "one": 1
}'''


def test_handle_type_error():
    @rbottle.handle_type_error
    def too_much_parameters(one):
        pass

    with pytest.raises(HTTPError):
        too_much_parameters("one", "two")

    @rbottle.handle_type_error
    def too_few_parameters(one, two, three):
        pass

    with pytest.raises(HTTPError):
        too_few_parameters("one")

    @rbottle.handle_type_error
    def exactly_right_ammount_of_parameters(one):
        pass

    assert exactly_right_ammount_of_parameters("one") is None


# json_to_params tests ========================================================
def test_json_to_params():
    rbottle.request = MockRequest('{"param": 2}')

    @rbottle.json_to_params
    def json_to_params_test(param):
        return param * 2

    assert json_to_params_test() == "4"  # serialized to json


def test_json_to_params_no_json_parameter():
    rbottle.request = MockRequest('{"param": 2}')

    @rbottle.json_to_params(return_json=False)
    def json_to_params_test_no_json(param):
        return param * 2

    assert json_to_params_test_no_json() == 4


def test_json_to_params_list():
    rbottle.request = MockRequest('[2]')  # different parameter is used

    @rbottle.json_to_params
    def json_to_params_test(param):
        return param * 2

    assert json_to_params_test() == "4"


def test_json_to_params_value():
    rbottle.request = MockRequest('2')  # different parameter is used

    @rbottle.json_to_params
    def json_to_params_test(param):
        return param * 2

    assert json_to_params_test() == "4"


def test_json_to_params_bad_keyword():
    rbottle.request = MockRequest('{"nope": 1}')

    @rbottle.json_to_params
    def json_to_params_test(param):
        return param * 2

    with pytest.raises(HTTPError):
        json_to_params_test()


# json_to_data tests ==========================================================
def test_json_to_data():
    rbottle.request = MockRequest('2')

    @rbottle.json_to_data
    def json_to_data_test(data):
        return data * 2

    assert json_to_data_test() == "4"


def test_json_to_data_no_json_parameter():
    rbottle.request = MockRequest('2')

    @rbottle.json_to_data(return_json=False)  # don't convert result to JSON
    def json_to_data_test(data):
        return data * 2

    assert json_to_data_test() == 4


def test_json_to_data_error_too_few():
    rbottle.request = MockRequest('2')

    @rbottle.json_to_data
    def json_to_data_test():
        pass

    with pytest.raises(HTTPError):
        json_to_data_test()


def test_json_to_data_error_too_many():
    rbottle.request = MockRequest('2')

    @rbottle.json_to_data
    def json_to_data_test(one, two):
        pass

    with pytest.raises(HTTPError):
        json_to_data_test()


# form_to_params tests ========================================================
def test_form_to_params():
    rbottle.request = MockRequest('""')
    rbottle.request.forms = {"param": 2}

    @rbottle.form_to_params
    def form_to_params_test(param):
        return param * 2

    assert form_to_params_test() == "4"


def test_form_to_params_no_json_parameter():
    rbottle.request = MockRequest('""')
    rbottle.request.forms = {"param": 2}

    @rbottle.form_to_params(return_json=False)  # don't convert result to JSON
    def form_to_params_test(param):
        return param * 2

    assert form_to_params_test() == 4


unittests/test_bottle_rest.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
import sys
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO

import pytest

sys.path = ['src/bottle_rest'] + sys.path
import bottle_rest
from bottle_rest import HTTPError


# Functions & classes =========================================================
class MockRequest:
    def __init__(self, json):
        self.body = StringIO(json)
        self.forms = {}


def test_decode_json_body():
    bottle_rest.request = MockRequest('{"hello": "there"}')

    assert bottle_rest.decode_json_body() == {"hello": "there"}

    with pytest.raises(HTTPError):
        bottle_rest.request = MockRequest('{')
        bottle_rest.decode_json_body()


def test_encode_json_body():
    assert bottle_rest.encode_json_body({"one": 1}) == '''{
    "one": 1
}'''


def test_handle_type_error():
    @bottle_rest.handle_type_error
    def too_much_parameters(one):
        pass

    with pytest.raises(HTTPError):
        too_much_parameters("one", "two")

    @bottle_rest.handle_type_error
    def too_few_parameters(one, two, three):
        pass

    with pytest.raises(HTTPError):
        too_few_parameters("one")

    @bottle_rest.handle_type_error
    def exactly_right_ammount_of_parameters(one):
        pass

    assert exactly_right_ammount_of_parameters("one") is None


# json_to_params tests ========================================================
def test_json_to_params():
    bottle_rest.request = MockRequest('{"param": 2}')

    @bottle_rest.json_to_params
    def json_to_params_test(param):
        return param * 2

    assert json_to_params_test() == "4"  # serialized to json


def test_json_to_params_no_json_parameter():
    bottle_rest.request = MockRequest('{"param": 2}')

    @bottle_rest.json_to_params(return_json=False)
    def json_to_params_test_no_json(param):
        return param * 2

    assert json_to_params_test_no_json() == 4


def test_json_to_params_list():
    bottle_rest.request = MockRequest('[2]')  # different parameter is used

    @bottle_rest.json_to_params
    def json_to_params_test(param):
        return param * 2

    assert json_to_params_test() == "4"


def test_json_to_params_value():
    bottle_rest.request = MockRequest('2')  # different parameter is used

    @bottle_rest.json_to_params
    def json_to_params_test(param):
        return param * 2

    assert json_to_params_test() == "4"


def test_json_to_params_bad_keyword():
    bottle_rest.request = MockRequest('{"nope": 1}')

    @bottle_rest.json_to_params
    def json_to_params_test(param):
        return param * 2

    with pytest.raises(HTTPError):
        json_to_params_test()


# json_to_data tests ==========================================================
def test_json_to_data():
    bottle_rest.request = MockRequest('2')

    @bottle_rest.json_to_data
    def json_to_data_test(data):
        return data * 2

    assert json_to_data_test() == "4"


def test_json_to_data_no_json_parameter():
    bottle_rest.request = MockRequest('2')

    @bottle_rest.json_to_data(return_json=False)  # don't convert result to JSON
    def json_to_data_test(data):
        return data * 2

    assert json_to_data_test() == 4


def test_json_to_data_error_too_few():
    bottle_rest.request = MockRequest('2')

    @bottle_rest.json_to_data
    def json_to_data_test():
        pass

    with pytest.raises(HTTPError):
        json_to_data_test()


def test_json_to_data_error_too_many():
    bottle_rest.request = MockRequest('2')

    @bottle_rest.json_to_data
    def json_to_data_test(one, two):
        pass

    with pytest.raises(HTTPError):
        json_to_data_test()


# form_to_params tests ========================================================
def test_form_to_params():
    bottle_rest.request = MockRequest('""')
    bottle_rest.request.forms = {"param": 2}

    @bottle_rest.form_to_params
    def form_to_params_test(param):
        return param * 2

    assert form_to_params_test() == "4"


def test_form_to_params_no_json_parameter():
    bottle_rest.request = MockRequest('""')
    bottle_rest.request.forms = {"param": 2}

    # don't convert result to JSON
    @bottle_rest.form_to_params(return_json=False)
    def form_to_params_test(param):
        return param * 2

    assert form_to_params_test() == 4


src/rbottle.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
import json
from functools import wraps

from bottle import request, HTTPError


# Functions & classes =========================================================
def decode_json_body():
    """
    Decode ``bottle.request.body`` to JSON.

    Returns:
        obj: Structure decoded by ``json.loads()``.

    Raises:
        HTTPError: 400 in case the data was malformed.
    """
    raw_data = request.body.read()

    try:
        return json.loads(raw_data)
    except ValueError as e:
        raise HTTPError(400, e.message)


def handle_type_error(fn):
    """
    Convert ``TypeError`` to ``bottle.HTTPError`` with ``400`` code and message
    about wrong parameters.

    Raises:
        HTTPError: 400 in case too many/too little function parameters were \
                   given.
    """
    def any_match(string_list, obj):
        return filter(lambda x: x in obj, string_list)

    @wraps(fn)
    def handle_type_error_wrapper(*args, **kwargs):
        try:
            return fn(*args, **kwargs)
        except TypeError as e:
            str_list = [
                "takes exactly",
                "got an unexpected",
                "takes no argument",
            ]
            if fn.__name__ in e.message and any_match(str_list, e.message):
                raise HTTPError(400, e.message)

            raise  # This will cause 500: Internal server error

    return handle_type_error_wrapper


def json_to_params(fn=None, return_json=True):
    """
    Convert JSON in the body of the request to the parameters for the wrapped
    function.

    If the JSON is list, add it to ``*args``.

    If dict, add it to ``**kwargs`` in non-rewrite mode (no key in ``**kwargs``
    will be overwritten).

    If single value, add it to ``*args``.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """
    def json_to_params_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def json_to_params_wrapper(*args, **kwargs):
            data = decode_json_body()

            if type(data) in [tuple, list]:
                args = list(args) + data
            elif type(data) == dict:
                # transport only items that are not already in kwargs
                allowed_keys = set(data.keys()) - set(kwargs.keys())
                for key in allowed_keys:
                    kwargs[key] = data[key]
            elif type(data) in [bool, int, float, long, str, unicode]:
                args = list(args)
                args.append(data)

            if not return_json:
                return fn(*args, **kwargs)

            return json.dumps(
                fn(*args, **kwargs)
            )

        return json_to_params_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return json_to_params_decorator(fn)

    return json_to_params_decorator


def json_to_data(fn=None, return_json=True):
    """
    Decode JSON from the request and add it as ``data`` parameter for wrapped
    function.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """
    def json_to_data_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def get_data_wrapper(*args, **kwargs):
            kwargs["data"] = decode_json_body()

            if not return_json:
                return fn(*args, **kwargs)

            return json.dumps(
                fn(*args, **kwargs)
            )

        return get_data_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return json_to_data_decorator(fn)

    return json_to_data_decorator


def form_to_params(fn=None, return_json=True):
    """
    Convert bottle forms request to parameters for the wrapped function.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """
    def forms_to_params_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def forms_to_params_wrapper(*args, **kwargs):
            kwargs.update(
                dict(request.forms)
            )

            if not return_json:
                return fn(*args, **kwargs)

            return json.dumps(
                fn(*args, **kwargs)
            )

        return forms_to_params_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return forms_to_params_decorator(fn)

    return forms_to_params_decorator


src/bottle_rest/bottle_rest.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
import json
from functools import wraps

from bottle import request, response, HTTPError


# Functions & classes =========================================================
def pretty_dump(fn):
    """
    Decorator used to output prettified JSON.

    ``response.content_type`` is set to ``application/json; charset=utf-8``.

    Args:
        fn (fn pointer): Function returning any basic python data structure.

    Returns:
        str: Data converted to prettified JSON.
    """
    @wraps(fn)
    def pretty_dump_wrapper(*args, **kwargs):
        response.content_type = "application/json; charset=utf-8"

        return json.dumps(
            fn(*args, **kwargs),

            # sort_keys=True,
            indent=4,
            separators=(', ', ': ')
        )

    return pretty_dump_wrapper


def decode_json_body():
    """
    Decode ``bottle.request.body`` to JSON.

    Returns:
        obj: Structure decoded by ``json.loads()``.

    Raises:
        HTTPError: 400 in case the data was malformed.
    """
    raw_data = request.body.read()

    try:
        return json.loads(raw_data)
    except ValueError as e:
        raise HTTPError(400, e.__str__())


def encode_json_body(data):
    """
    Return prettified JSON `data`, set ``response.content_type`` to
    ``application/json; charset=utf-8``.

    Args:
        data (any): Any basic python data structure.

    Returns:
        str: Data converted to prettified JSON.
    """
    response.content_type = "application/json; charset=utf-8"

    return json.dumps(
        data,
        indent=4,
        separators=(',', ': ')
    )


def handle_type_error(fn):
    """
    Convert ``TypeError`` to ``bottle.HTTPError`` with ``400`` code and message
    about wrong parameters.

    Raises:
        HTTPError: 400 in case too many/too little function parameters were \
                   given.
    """
    @wraps(fn)
    def handle_type_error_wrapper(*args, **kwargs):
        def any_match(string_list, obj):
            return filter(lambda x: x in obj, string_list)

        try:
            return fn(*args, **kwargs)
        except TypeError as e:
            message = e.__str__()
            str_list = [
                "takes exactly",
                "got an unexpected",
                "takes no argument",
            ]
            if fn.__name__ in message and any_match(str_list, message):
                raise HTTPError(400, message)

            raise  # This will cause 500: Internal server error

    return handle_type_error_wrapper


def json_to_params(fn=None, return_json=True):
    """
    Convert JSON in the body of the request to the parameters for the wrapped
    function.

    If the JSON is list, add it to ``*args``.

    If dict, add it to ``**kwargs`` in non-rewrite mode (no key in ``**kwargs``
    will be overwritten).

    If single value, add it to ``*args``.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """
    def json_to_params_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def json_to_params_wrapper(*args, **kwargs):
            data = decode_json_body()

            if type(data) in [tuple, list]:
                args = list(args) + data
            elif type(data) == dict:
                # transport only items that are not already in kwargs
                allowed_keys = set(data.keys()) - set(kwargs.keys())
                for key in allowed_keys:
                    kwargs[key] = data[key]
            elif type(data) in [bool, int, float, long, str, unicode]:
                args = list(args)
                args.append(data)

            if not return_json:
                return fn(*args, **kwargs)

            return encode_json_body(
                fn(*args, **kwargs)
            )

        return json_to_params_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return json_to_params_decorator(fn)

    return json_to_params_decorator


def json_to_data(fn=None, return_json=True):
    """
    Decode JSON from the request and add it as ``data`` parameter for wrapped
    function.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """
    def json_to_data_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def get_data_wrapper(*args, **kwargs):
            kwargs["data"] = decode_json_body()

            if not return_json:
                return fn(*args, **kwargs)

            return encode_json_body(
                fn(*args, **kwargs)
            )

        return get_data_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return json_to_data_decorator(fn)

    return json_to_data_decorator


def form_to_params(fn=None, return_json=True):
    """
    Convert bottle forms request to parameters for the wrapped function.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """
    def forms_to_params_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def forms_to_params_wrapper(*args, **kwargs):
            kwargs.update(
                dict(request.forms)
            )

            if not return_json:
                return fn(*args, **kwargs)

            return encode_json_body(
                fn(*args, **kwargs)
            )

        return forms_to_params_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return forms_to_params_decorator(fn)

    return forms_to_params_decorator


src/bottle_rest/__init__.py
from json import dumps
from json import loads
from functools import wraps

from bottle import request
from bottle import response
from bottle import HTTPError


PRIMITIVE_TYPES = {bool, int, float, str}


def pretty_dump(fn):
    """
    Decorator used to output prettified JSON.

    ``response.content_type`` is set to ``application/json; charset=utf-8``.

    Args:
        fn (fn pointer): Function returning any basic python data structure.

    Returns:
        str: Data converted to prettified JSON.
    """

    @wraps(fn)
    def pretty_dump_wrapper(*args, **kwargs):
        response.content_type = "application/json; charset=utf-8"

        return dumps(fn(*args, **kwargs), indent=4, separators=(",", ": "))

    return pretty_dump_wrapper


def decode_json_body():
    """
    Decode ``bottle.request.body`` to JSON.

    Returns:
        obj: Structure decoded by ``json.loads()``.

    Raises:
        HTTPError: 400 in case the data was malformed.
    """
    raw_data = request.body.read() or "{}"

    try:
        return loads(raw_data)
    except ValueError as e:
        raise HTTPError(400, e.__str__())


def encode_json_body(data):
    """
    Return prettified JSON `data`, set ``response.content_type`` to
    ``application/json; charset=utf-8``.

    Args:
        data (any): Any basic python data structure.

    Returns:
        str: Data converted to prettified JSON.
    """
    # support for StringIO / file - like objects
    if hasattr(data, "read"):
        return data

    response.content_type = "application/json; charset=utf-8"

    return dumps(data, indent=4, separators=(",", ": "))


def handle_type_error(fn):
    """
    Convert ``TypeError`` to ``bottle.HTTPError`` with ``400`` code and message
    about wrong parameters.

    Raises:
        HTTPError: 400 in case too many/too little function parameters were \
                   given.
    """

    @wraps(fn)
    def handle_type_error_wrapper(*args, **kwargs):
        def any_match(string_list, obj):
            return filter(lambda x: x in obj, string_list)

        try:
            return fn(*args, **kwargs)
        except TypeError as e:
            message = e.__str__()
            str_list = [
                "takes exactly",
                "got an unexpected",
                "takes no argument",
            ]
            if fn.__name__ in message and any_match(str_list, message):
                raise HTTPError(400, message)

            raise  # This will cause 500: Internal server error

    return handle_type_error_wrapper


def json_to_params(fn=None, return_json=True):
    """
    Convert JSON in the body of the request to the parameters for the wrapped
    function.

    If the JSON is list, add it to ``*args``.

    If dict, add it to ``**kwargs`` in non-rewrite mode (no key in ``**kwargs``
    will be overwritten).

    If single value, add it to ``*args``.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """

    def json_to_params_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def json_to_params_wrapper(*args, **kwargs):
            data = decode_json_body()

            if type(data) in [tuple, list]:
                args = list(args) + data
            elif type(data) == dict:
                # transport only items that are not already in kwargs
                allowed_keys = set(data.keys()) - set(kwargs.keys())
                for key in allowed_keys:
                    kwargs[key] = data[key]
            elif type(data) in PRIMITIVE_TYPES:
                args = list(args)
                args.append(data)

            if not return_json:
                return fn(*args, **kwargs)

            return encode_json_body(fn(*args, **kwargs))

        return json_to_params_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return json_to_params_decorator(fn)

    return json_to_params_decorator


def json_to_data(fn=None, return_json=True):
    """
    Decode JSON from the request and add it as ``data`` parameter for wrapped
    function.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """

    def json_to_data_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def get_data_wrapper(*args, **kwargs):
            kwargs["data"] = decode_json_body()

            if not return_json:
                return fn(*args, **kwargs)

            return encode_json_body(fn(*args, **kwargs))

        return get_data_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return json_to_data_decorator(fn)

    return json_to_data_decorator


def form_to_params(fn=None, return_json=True):
    """
    Convert bottle forms request to parameters for the wrapped function.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """

    def forms_to_params_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def forms_to_params_wrapper(*args, **kwargs):
            kwargs.update(dict(request.forms))

            if not return_json:
                return fn(*args, **kwargs)

            return encode_json_body(fn(*args, **kwargs))

        return forms_to_params_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return forms_to_params_decorator(fn)

    return forms_to_params_decorator


def get_to_params(fn=None, return_json=True):
    """
    Convert bottle get parameters to parameters for the wrapped function.

    Some of the JavaScript libraries don't work well with sending the
    JSON body via GET request, so you can use this.

    Args:
        return_json (bool, default True): Should the decorator automatically
                    convert returned value to JSON?
    """

    def get_to_params_decorator(fn):
        @handle_type_error
        @wraps(fn)
        def forms_to_params_wrapper(*args, **kwargs):
            kwargs.update(dict(request.query.decode()))

            if not return_json:
                return fn(*args, **kwargs)

            return encode_json_body(fn(*args, **kwargs))

        return forms_to_params_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return get_to_params_decorator(fn)

    return get_to_params_decorator
